### PR TITLE
Fixed the auto-scroll right bug in a scaled browser.

### PR DESCRIPTION
**Pull request by cw196**
**Issue Number**

[`#528`](https://github.com/namespace-ee/react-calendar-timeline/issues/528)

**Overview of PR**

In a scaled the browser, for example to 75%, the screen keeps moving to the left in some case. This issue is caused by the value passed back from`onscroll` callback is decimal in a scaled browser. Then when we `// Check the scroll is correct` in componenDidUpdate, we compare that decimal number with an integer value, which is always not equal, and manually 'scroll', then causes an infinite loop.

The fix is that wrapping the 'to be compared `scrollLeft` variable'  with a Math.round(), which prevent the loop when a decimal value is set to `scrollLeft`.

Please re-edit the file 'demo/app/demo-main/index.js' with the following code. And the 'auto-scroll' bug can be triggered in a scaled browser (if you cannot reproduce, please try a different scaled size or resize the scaled browser.) with the original code. 

```
/* eslint-disable no-console */
import React, { Component } from 'react'
import moment from 'moment'

import debounce from 'lodash/debounce';


import Timeline, {
  TimelineMarkers,
  TimelineHeaders,
  TodayMarker,
  CustomMarker,
  CursorMarker,
  CustomHeader,
  SidebarHeader,
  DateHeader
} from 'react-calendar-timeline'

import generateFakeData, { fakeItem } from '../generate-fake-data'

var minTime = moment()
  .add(-6, 'months')
  .valueOf()
var maxTime = moment()
  .add(6, 'months')
  .valueOf()

var keys = {
  groupIdKey: 'id',
  groupTitleKey: 'title',
  groupRightTitleKey: 'rightTitle',
  itemIdKey: 'id',
  itemTitleKey: 'title',
  itemDivTitleKey: 'title',
  itemGroupKey: 'group',
  itemTimeStartKey: 'start',
  itemTimeEndKey: 'end'
}

export default class App extends Component {
  constructor(props) {
    super(props)

    const { groups, items } = generateFakeData()

    const visibleTimeStart = moment()
      .startOf('day')
      .toDate().valueOf();
    const visibleTimeEnd = moment()
      .startOf('day')
      .add(1, 'day')
      .toDate().valueOf();

    this.state = {
      groups,
      items,
      visibleTimeStart,
      visibleTimeEnd
    }
  }

  handleCanvasClick = (groupId, time) => {
    console.log('Canvas clicked', groupId, moment(time).format())
  }

  handleCanvasDoubleClick = (groupId, time) => {
    console.log('Canvas double clicked', groupId, moment(time).format())
  }

  handleCanvasContextMenu = (group, time) => {
    console.log('Canvas context menu', group, moment(time).format())
  }

  handleItemClick = (itemId, _, time) => {
    console.log('Clicked: ' + itemId, moment(time).format())
  }

  handleItemSelect = (itemId, _, time) => {
    console.log('Selected: ' + itemId, moment(time).format())
  }

  handleItemDoubleClick = (itemId, _, time) => {
    console.log('Double Click: ' + itemId, moment(time).format())
  }

  handleItemContextMenu = (itemId, _, time) => {
    console.log('Context Menu: ' + itemId, moment(time).format())
  }

  handleItemMove = (itemId, dragTime, newGroupOrder) => {
    const { items, groups } = this.state

    const group = groups[newGroupOrder]

    this.setState({
      items: items.map(item =>
        item.id === itemId
          ? Object.assign({}, item, {
            start: dragTime,
            end: dragTime + (item.end - item.start),
            group: group.id
          })
          : item
      )
    })

    console.log('Moved', itemId, dragTime, newGroupOrder)
  }

  handleItemResize = (itemId, time, edge) => {
    const { items } = this.state

    this.setState({
      items: items.map(item =>
        item.id === itemId
          ? Object.assign({}, item, {
            start: edge === 'left' ? time : item.start,
            end: edge === 'left' ? item.end : time
          })
          : item
      )
    })

    console.log('Resized', itemId, time, edge)
  }

  handleTimeChange = (visibleTimeStart, visibleTimeEnd, updateScrollCanvas) => {
    updateScrollCanvas(visibleTimeStart, visibleTimeEnd)
    this.setState({
      visibleTimeStart,
      visibleTimeEnd
    })
  }

  componentDidUpdate(preProps, preState) {
    if (this.state.visibleTimeStart !== preState.visibleTimeStart) {
      // imagine we are calling Backend API.
      this.fetchAPIDate();
    }
  }


  fetchAPIDate = debounce(
    () =>
      this.setState({
        items: generateFakeData().items
      }),
    2000,
  );


  moveResizeValidator = (action, item, time) => {
    if (time < new Date().getTime()) {
      var newTime =
        Math.ceil(new Date().getTime() / (15 * 60 * 1000)) * (15 * 60 * 1000)
      return newTime
    }

    return time
  }

  render() {
    const { groups, items, visibleTimeStart, visibleTimeEnd } = this.state

    return (
      <Timeline
        groups={groups}
        items={items}
        keys={keys}
        sidebarWidth={150}
        sidebarContent={<div>Above The Left</div>}
        canMove
        canResize="right"
        canSelect
        itemsSorted
        itemTouchSendsClick={false}
        stackItems
        itemHeightRatio={0.75}
        visibleTimeStart={visibleTimeStart}
        visibleTimeEnd={visibleTimeEnd}
        onCanvasClick={this.handleCanvasClick}
        onCanvasDoubleClick={this.handleCanvasDoubleClick}
        onCanvasContextMenu={this.handleCanvasContextMenu}
        onItemClick={this.handleItemClick}
        onItemSelect={this.handleItemSelect}
        onItemContextMenu={this.handleItemContextMenu}
        onItemMove={this.handleItemMove}
        onItemResize={this.handleItemResize}
        onItemDoubleClick={this.handleItemDoubleClick}
        onTimeChange={this.handleTimeChange}
        moveResizeValidator={this.moveResizeValidator}
      >
        <TimelineMarkers>
          <TodayMarker />
          <CustomMarker
            date={
              moment()
                .startOf('day')
                .valueOf() +
              1000 * 60 * 60 * 2
            }
          />
          <CustomMarker
            date={moment()
              .add(3, 'day')
              .valueOf()}
          >
            {({ styles }) => {
              const newStyles = { ...styles, backgroundColor: 'blue' }
              return <div style={newStyles} />
            }}
          </CustomMarker>
          <CursorMarker />
        </TimelineMarkers>
      </Timeline>
    )
  }
}
```
**Demo**
### Before
old
### Now
new

### CODE REVIEW

**Comment by Ilaiwi** [comment]
@cw196 Thank you for the PR. I really appreciate your contribution. I will test this and release it soon.

**Comment by cw196** [comment]
Thank you for your work @Ilaiwi 🙂

File: src/lib/Timeline.js [hunk 1]
```diff
@@ -425,11 +425,12 @@ export default class ReactCalendarTimeline extends Component {
...
         newZoom
     )
-    if (this.scrollComponent.scrollLeft !== scrollLeft) {
+    const componentScrollLeft = Math.round(this.scrollComponent.scrollLeft);
+    if (componentScrollLeft !== scrollLeft) {
       this.scrollComponent.scrollLeft = scrollLeft
     }
 
-    if (this.scrollHeaderRef.scrollLeft !== scrollLeft) {
+    if (componentScrollLeft !== scrollLeft) {
```
Comment (tobiasbueschel) [inline src/lib/Timeline.js]: Thanks a lot for this PR, I've been having the same issues!

If we are doing the same conditional if check here `componentScrollLeft !== scrollLeft`, we could also move `this.scrollHeaderRef.scrollLeft = scrollLeft` into the previous `if` block :-)

Comment (cw196) [reply src/lib/Timeline.js]: Yeah! Done! @tobiasbueschel

**Comment by cw196** [comment]
Do you guys know when we could publish the fix? @Ilaiwi @tobiasbueschel .
Thanks so much for your hard work, guys!

**Pull request merged**